config/config.go

```diff
@@ -244,12 +245,19 @@ func loadExtensions(v *viper.Viper, factories map[string]extension.Factory) (con
 		extensionCfg.SetType(typeStr)
 		extensionCfg.SetName(fullName)
 
-		// Now that the default config struct is created we can Unmarshal into it
-		// and it will apply user-defined config on top of the default.
-		if err := subViper.UnmarshalKey(key, extensionCfg); err != nil {
-			return nil, &configError{
-				code: errUnmarshalError,
-				msg:  fmt.Sprintf("error reading settings for extension type %q: %v", typeStr, err),
+		// Unmarsh only the subconfig for this extension.
+		sv := subViper.Sub(key)
+		if sv != nil {
+			// Before unmarshilng first expand all environment variables.
+			sv = expandEnvConfig(sv)
```

Comment:
I guess that you hit some issue when this was initially on `Load` can you add a small comment to clarify that? Otherwise later someone is bound to attempt to move this to `Load`.

Reply:
Actually the biggest concern was that, I want this functionality only for the components config not for other config like allowing to add a full new exporter.